File: src/api/ga4_client.py
"""Google Analytics 4 Data API client."""

import pandas as pd
from datetime import datetime, timedelta
from typing import List, Optional
from google.analytics.data_v1beta.types import (
    RunReportRequest,
    DateRange,
    Dimension,
    Metric,
    OrderBy
)
from tenacity import retry, stop_after_attempt, wait_exponential

from src.auth.google_auth import get_ga4_client
from src.config.settings import settings, SiteConfig


class GA4Client:
    """Client for interacting with GA4 Data API."""

    def __init__(self, site_config: Optional[SiteConfig] = None):
        """
        Initialize GA4 client.

        Args:
            site_config: Optional site configuration. If not provided,
                        falls back to legacy single-site settings.
        """
        self.client = get_ga4_client()

        if site_config:
            self.property_id = site_config.ga4_property_id
        else:
            self.property_id = settings.ga4_property_id

    @property
    def property_path(self) -> str:
        """Get the GA4 property path."""
        return f"properties/{self.property_id}"

    @retry(
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=1, min=2, max=10)
    )
    def run_report(
        self,
        dimensions: List[str],
        metrics: List[str],
        start_date: str,
        end_date: str,
        order_by: Optional[str] = None,
        limit: int = 10000
    ) -> pd.DataFrame:
        """
        Execute a GA4 report and return as DataFrame.

        Args:
            dimensions: List of dimension names (e.g., ['date', 'pagePath'])
            metrics: List of metric names (e.g., ['activeUsers', 'sessions'])
            start_date: Start date in YYYY-MM-DD format
            end_date: End date in YYYY-MM-DD format
            order_by: Optional metric name to order by (descending)
            limit: Maximum rows to return

        Returns:
            pd.DataFrame: Report data
        """
        request = RunReportRequest(
            property=self.property_path,
            dimensions=[Dimension(name=d) for d in dimensions],
            metrics=[Metric(name=m) for m in metrics],
            date_ranges=[DateRange(start_date=start_date, end_date=end_date)],
            limit=limit
        )

        if order_by:
            request.order_bys = [
                OrderBy(
                    metric=OrderBy.MetricOrderBy(metric_name=order_by),
                    desc=True
                )
            ]

        response = self.client.run_report(request)

        # Convert to DataFrame
        rows = []
        for row in response.rows:
            row_data = {}
            for i, dim in enumerate(dimensions):
                row_data[dim] = row.dimension_values[i].value
            for i, met in enumerate(metrics):
                value = row.metric_values[i].value
                # Try to convert to numeric
                try:
                    row_data[met] = float(value)
                except ValueError:
                    row_data[met] = value
            rows.append(row_data)

        df = pd.DataFrame(rows)

        # Convert date column if present
        if 'date' in df.columns:
            df['date'] = pd.to_datetime(df['date'], format='%Y%m%d')

        return df

    def get_traffic_overview(self, start_date: str, end_date: str) -> dict:
        """
        Get key traffic metrics for the date range.

        Returns:
            dict: Traffic overview with totals and trend data
        """
        df = self.run_report(
            dimensions=['date'],
            metrics=[
                'activeUsers',
                'sessions',
                'screenPageViews',
                'bounceRate',
                'averageSessionDuration'
            ],
            start_date=start_date,
            end_date=end_date
        )

        if df.empty:
            return {
                'users': 0,
                'sessions': 0,
                'pageviews': 0,
                'bounce_rate': 0,
                'avg_duration': 0,
                'trend_data': df
            }

        return {
            'users': int(df['activeUsers'].sum()),
            'sessions': int(df['sessions'].sum()),
            'pageviews': int(df['screenPageViews'].sum()),
            'bounce_rate': round(df['bounceRate'].mean() * 100, 2),
            'avg_duration': round(df['averageSessionDuration'].mean(), 1),
            'trend_data': df.sort_values('date')
        }

    def get_traffic_sources(self, start_date: str, end_date: str) -> pd.DataFrame:
        """Get traffic breakdown by channel."""
        return self.run_report(
            dimensions=['sessionDefaultChannelGrouping'],
            metrics=['sessions', 'activeUsers', 'bounceRate'],
            start_date=start_date,
            end_date=end_date,
            order_by='sessions'
        )

    def get_top_pages(
        self,
        start_date: str,
        end_date: str,
        limit: int = 20
    ) -> pd.DataFrame:
        """Get top pages by pageviews."""
        return self.run_report(
            dimensions=['pagePath'],
            metrics=['screenPageViews', 'activeUsers', 'averageSessionDuration'],
            start_date=start_date,
            end_date=end_date,
            order_by='screenPageViews',
            limit=limit
        )

    def get_devices(self, start_date: str, end_date: str) -> pd.DataFrame:
        """Get traffic by device category."""
        return self.run_report(
            dimensions=['deviceCategory'],
            metrics=['sessions', 'activeUsers'],
            start_date=start_date,
            end_date=end_date,
            order_by='sessions'
        )

    def get_browsers(self, start_date: str, end_date: str) -> pd.DataFrame:
        """Get traffic by browser."""
        return self.run_report(
            dimensions=['browser'],
            metrics=['sessions', 'activeUsers'],
            start_date=start_date,
            end_date=end_date,
            order_by='sessions',
            limit=10
        )

    def get_countries(
        self,
        start_date: str,
        end_date: str,
        limit: int = 10
    ) -> pd.DataFrame:
        """Get traffic by country."""
        return self.run_report(
            dimensions=['country'],
            metrics=['sessions', 'activeUsers'],
            start_date=start_date,
            end_date=end_date,
            order_by='sessions',
            limit=limit
        )

    def get_events(self, start_date: str, end_date: str) -> pd.DataFrame:
        """Get event counts by event name."""
        return self.run_report(
            dimensions=['eventName'],
            metrics=['eventCount', 'eventValue'],
            start_date=start_date,
            end_date=end_date,
            order_by='eventCount'
        )

    def get_events_trend(self, start_date: str, end_date: str) -> pd.DataFrame:
        """Get event counts over time."""
        return self.run_report(
            dimensions=['date'],
            metrics=['eventCount', 'conversions'],
            start_date=start_date,
            end_date=end_date
        )

    def get_conversions(self, start_date: str, end_date: str) -> pd.DataFrame:
        """Get conversion events."""
        return self.run_report(
            dimensions=['eventName'],
            metrics=['conversions', 'eventValue'],
            start_date=start_date,
            end_date=end_date,
            order_by='conversions'
        )

File: dashboard/src/lib/ga4-client.ts
/**
 * GA4 Data API client.
 */

import { getGA4Client } from './google-auth';
import { SiteConfig } from './settings';

export interface TrafficOverview {
  users: number;
  sessions: number;
  pageviews: number;
  bounceRate: number;
  avgDuration: number;
  trendData: Array<{
    date: string;
    activeUsers: number;
    sessions: number;
    screenPageViews: number;
    bounceRate: number;
    averageSessionDuration: number;
  }>;
}

export class GA4DataClient {
  private propertyId: string;
  private client: ReturnType<typeof getGA4Client>;

  constructor(siteConfig: SiteConfig) {
    this.propertyId = siteConfig.ga4PropertyId;
    this.client = getGA4Client();
  }

  private get propertyPath(): string {
    return `properties/${this.propertyId}`;
  }

  /**
   * Run a GA4 report.
   */
  async runReport(params: {
    dimensions: string[];
    metrics: string[];
    startDate: string;
    endDate: string;
    orderBy?: string;
    limit?: number;
  }) {
    const requestBody: Record<string, unknown> = {
      dimensions: params.dimensions.map(name => ({ name })),
      metrics: params.metrics.map(name => ({ name })),
      dateRanges: [{ startDate: params.startDate, endDate: params.endDate }],
      limit: String(params.limit || 10000),
    };

    if (params.orderBy) {
      requestBody.orderBys = [{ metric: { metricName: params.orderBy }, desc: true }];
    }

    const response = await this.client.properties.runReport({
      property: this.propertyPath,
      requestBody,
    });

    const rows: Record<string, any>[] = [];
    for (const row of response.data.rows || []) {
      const rowData: Record<string, any> = {};
      params.dimensions.forEach((dim, i) => {
        rowData[dim] = row.dimensionValues?.[i]?.value || '';
      });
      params.metrics.forEach((met, i) => {
        const value = row.metricValues?.[i]?.value || '0';
        rowData[met] = parseFloat(value) || 0;
      });
      rows.push(rowData);
    }

    return rows;
  }

  /**
   * Get traffic overview for date range.
   */
  async getTrafficOverview(startDate: string, endDate: string): Promise<TrafficOverview> {
    const rows = await this.runReport({
      dimensions: ['date'],
      metrics: [
        'activeUsers',
        'sessions',
        'screenPageViews',
        'bounceRate',
        'averageSessionDuration',
      ],
      startDate,
      endDate,
    });

    if (rows.length === 0) {
      return {
        users: 0,
        sessions: 0,
        pageviews: 0,
        bounceRate: 0,
        avgDuration: 0,
        trendData: [],
      };
    }

    const totalUsers = rows.reduce((sum, r) => sum + r.activeUsers, 0);
    const totalSessions = rows.reduce((sum, r) => sum + r.sessions, 0);
    const totalPageviews = rows.reduce((sum, r) => sum + r.screenPageViews, 0);
    const avgBounceRate = rows.reduce((sum, r) => sum + r.bounceRate, 0) / rows.length;
    const avgDuration = rows.reduce((sum, r) => sum + r.averageSessionDuration, 0) / rows.length;

    // Sort by date
    const trendData = rows
      .map(r => ({
        date: r.date,
        activeUsers: r.activeUsers,
        sessions: r.sessions,
        screenPageViews: r.screenPageViews,
        bounceRate: r.bounceRate,
        averageSessionDuration: r.averageSessionDuration,
      }))
      .sort((a, b) => a.date.localeCompare(b.date));

    return {
      users: Math.round(totalUsers),
      sessions: Math.round(totalSessions),
      pageviews: Math.round(totalPageviews),
      bounceRate: Math.round(avgBounceRate * 100 * 100) / 100, // percentage with 2 decimals
      avgDuration: Math.round(avgDuration * 10) / 10,
      trendData,
    };
  }

  /**
   * Get traffic sources breakdown.
   */
  async getTrafficSources(startDate: string, endDate: string) {
    return this.runReport({
      dimensions: ['sessionDefaultChannelGrouping'],
      metrics: ['sessions', 'activeUsers', 'bounceRate'],
      startDate,
      endDate,
      orderBy: 'sessions',
    });
  }

  /**
   * Get top pages by pageviews.
   */
  async getTopPages(startDate: string, endDate: string, limit = 20) {
    return this.runReport({
      dimensions: ['pagePath'],
      metrics: ['screenPageViews', 'activeUsers', 'averageSessionDuration'],
      startDate,
      endDate,
      orderBy: 'screenPageViews',
      limit,
    });
  }

  /**
   * Get device breakdown.
   */
  async getDevices(startDate: string, endDate: string) {
    return this.runReport({
      dimensions: ['deviceCategory'],
      metrics: ['sessions', 'activeUsers'],
      startDate,
      endDate,
      orderBy: 'sessions',
    });
  }

  /**
   * Get country breakdown.
   */
  async getCountries(startDate: string, endDate: string, limit = 10) {
    return this.runReport({
      dimensions: ['country'],
      metrics: ['sessions', 'activeUsers'],
      startDate,
      endDate,
      orderBy: 'sessions',
      limit,
    });
  }

  /**
   * Get browser breakdown.
   */
  async getBrowsers(startDate: string, endDate: string, limit = 10) {
    return this.runReport({
      dimensions: ['browser'],
      metrics: ['sessions', 'activeUsers'],
      startDate,
      endDate,
      orderBy: 'sessions',
      limit,
    });
  }

  /**
   * Get events breakdown by name.
   */
  async getEvents(startDate: string, endDate: string, limit = 50) {
    return this.runReport({
      dimensions: ['eventName'],
      metrics: ['eventCount', 'eventValue'],
      startDate,
      endDate,
      orderBy: 'eventCount',
      limit,
    });
  }

  /**
   * Get events trend over time.
   */
  async getEventsTrend(startDate: string, endDate: string) {
    return this.runReport({
      dimensions: ['date'],
      metrics: ['eventCount', 'conversions'],
      startDate,
      endDate,
    });
  }

  /**
   * Get conversion events.
   */
  async getConversions(startDate: string, endDate: string, limit = 20) {
    return this.runReport({
      dimensions: ['eventName'],
      metrics: ['conversions', 'eventValue'],
      startDate,
      endDate,
      orderBy: 'conversions',
      limit,
    });
  }
}
